### PR TITLE
lib/osutil: Replace IsDir with TraversesSymlink (fixes #3839)

### LINKED ISSUE #3839: Resync of offline machine stuck in "Out of Sync" state after dirs+files were added and then removed elsewhere
## Current Behavior
I am seeing syncthing get into a blocked "Out of Sync" state every time I switch branches in a code repo I have synchronized. This is a repo that I have been syncing for 6+ months without issue, but is now (with 0.14.16) failing. I cannot say for certain if 0.14.16 is the cause or if it was 0.14.14 or 15 as I haven't done a checkout of this particular branch for about a week prior to today. 

I currently see 27 items (a mix of files and directories) "unable to sync" due to a missing parent directory. Here is a sample screenshot from the UI with filenames scrambled.

screenshot from 2016-12-23 18-34-47

Looking at the paths the directories are all sub-directories of a (missing) parent directory, and the files are contained in those sub-directories, so that might be an indication of where the issue is. Also of note is that all of the indicated files and directories are not currently present on either machine. I had switched to a branch, then switched back while the failing machine was offline. The syncing machine state _should_ be a no-op actually because it's state matches that of the remote. I should note that I experienced the exact same thing on another machine and "fixed" it by switching branches directly on that machine.

## Expected Behavior
- Syncthing completes sync of all non-existent directories once offline machine is reconnected

I will try to get a reproduction script together, but it may take a little bit of time to pull off. Happy to run any suggested diagnostics on the failing machine. I seem to be running into all the syncthing bugs lately after 6+ months of worry-free use :-/

### Version Information

Syncthing Version: v0.14.16
OS Version: Arch Linux (x64)
Browser Version: N/A

### PR DESCRIPTION
**Pull request by AudriusButkevicius**
This is an alternative, which just checks for symlink traversal, which was our main security issue

/cc @Unrud

### CODE REVIEW

**Comment by AudriusButkevicius** [comment]
On android this might cause a bit of stupidness, as EvalSymlinks is not using our magical lstat, but it's probably fine.

**Comment by AudriusButkevicius** [comment]
I guess we could fix Lstat with:

https://github.com/bouk/monkey

**Comment by calmh** [comment]
I'm skeptical as EvalSymlinks does a lot of work that we don't need, much more than our simple lstat loop. It's also checking for a negative rather then ensuring a positive (i.e., a black list rather than a white list). I think your other approach to verify that everything along the way is a directory (or potentially nothing at all) is better.

**Comment by AudriusButkevicius** [comment]
But this better represents what exactly we are trying to catch, and I wouldn't say its doing much more work.

**Comment by calmh** [comment]
Fair enough, it's not exactly as bad as I was afraid of;

```
BenchmarkIsDir-8              	  200000	     11022 ns/op	    2096 B/op	      21 allocs/op
BenchmarkTraversesSymlink-8   	  100000	     15993 ns/op	    2784 B/op	      35 allocs/op
```

**Comment by AudriusButkevicius** [comment]
So this renames stuff, yet keeps the functionality as discussed.

**Comment by AudriusButkevicius** [comment]
Anyways, both PRs now fix the issue. Tested manually with the repro provided. Merge whichever one you want.

File: lib/osutil/traversessymlink_test.go [hunk 1]
```diff
@@ -35,40 +35,42 @@ func TestIsDir(t *testing.T) {
...
 		{"a/x", false},
 		{"a/b/x", false},
 		{"a/x/c", false},
 		// Symlink or behind symlink
-		{"a/l", false},
-		{"a/l/c", false},
+		{"a/l", true},
+		{"a/l/c", true},
+		// Non-existing behind a symlink, but does not fail as does not exist.
+		{"a/l/x", false},
```
Comment (calmh) [inline lib/osutil/traversessymlink_test.go]: But it should fail, or we will attempt `mkdir("a/l/x")` which we should not.

**Comment by calmh** [comment]
Right, you said as much in your comment above, sorry.

**Pull request closed without merging**